충남대 FE 정지원 - 2단계 주문하기 페이지 구현하기

jjw5655 opened the pull request:
커밋 메시지와 readme참고 부탁드려요.
이번 리뷰도 잘 부탁드립니다!

jjw5655 committed:
feat: AuthContext와 Provider 구현으로 전역 로그인 상태 관리

- User, AuthContextType 타입 정의
- AuthContext와 AuthProvider 구현
- useAuth 커스텀 훅 생성으로 어디서든 로그인 상태 접근 가능
- main.tsx에서 AuthProvider로 앱 전체 감싸기
- 타입 및 훅 export 설정 완료

jjw5655 committed:
feat: localStorage를 활용한 로그인 상태 영구 보존

- storage.ts 유틸리티 함수 구현 (저장/복원/삭제)
- AuthProvider에 초기화 로직 추가 (useEffect)
- 로그인 성공 시 localStorage 저장
- 로그아웃 시 localStorage 정리
- 새로고침해도 로그인 상태 유지됨

jjw5655 committed:
feat: 로그인 상태별 프로필 버튼 동작 분기 구현

- LoginPage와 AuthContext 연동 완료
- 로그인 성공 시 전역 상태 업데이트
- 프로필 버튼 클릭 시 로그인 상태에 따라 분기
  - 로그인: /my (마이페이지) 이동
  - 비로그인: /login (로그인) 이동
- 기본 MyPage 컴포넌트 및 라우팅 추가
- 로딩 상태 UI 반영

jjw5655 committed:
feat: PrivateRoute 컴포넌트로 마이페이지 접근 보호

- PrivateRoute 컴포넌트 구현
  - 로그인 상태 확인 및 로딩 처리
  - 비로그인 시 로그인 페이지로 리다이렉트
  - 현재 URL 정보 유지하여 로그인 후 원래 페이지로 복귀
- MyPage를 PrivateRoute로 보호
- 인증이 필요한 페이지 보안 강화

jjw5655 committed:
feat: 마이페이지에 로그아웃 기능 추가

- 마이페이지에서 환영 메시지와 이메일 표시
- 로그아웃 버튼 추가 및 클릭 시 로그아웃 처리 후 /login으로 이동

jjw5655 committed:
feat: 주문하기용 카드 템플릿 목데이터(cardTemplates) 추가

- src/data/cardTemplates.ts에 CardTemplate 타입 및 카드 템플릿 배열 추가
- 주문 메시지 카드 선택 기능 구현에 활용 가능

jjw5655 committed:
feat: 상품 클릭 시 주문하기 페이지 이동 및 주문 UI/UX 구현

- 상품 클릭 시 /order/:productId로 이동
- 주문하기 페이지 전체 UI 구현 (카드 템플릿 슬라이더, 메시지, 보내는/받는 사람, 상품 정보, 하단 주문 버튼)
  - 카드 템플릿 mock 데이터 기반 선택 및 미리보기 기능
  - 메시지, 이름, 전화번호, 수량 입력 폼 구현
  - 휴대폰 번호 자동 포맷팅 및 유효성 검사(010-xxxx-xxxx)
  - 잘못된 전화번호 입력 시 에러 메시지 표시

jjw5655 committed:
feat: 주문하기 페이지 접근 시 비로그인 상태면 로그인 유도 및 리다이렉트 처리

- /order/:productId 라우트에 PrivateRoute 적용
- 비로그인 사용자가 주문 페이지 접근 시 로그인 페이지로 이동
- 로그인 성공 시 원래 가려던 주문 페이지로 자동 이동 (state.from 활용)

jjw5655 committed:
feat: 주문 폼 유효성 검사 및 주문 완료 안내/리다이렉트 구현

- 메시지, 보내는 사람, 받는 사람, 전화번호, 수량 필드 유효성 검사 및 에러 메시지 표시
- 전화번호 입력은 010-xxxx-xxxx 형식만 허용
- 모든 정보가 올바르면 주문 완료 안내 메시지(alert) 출력
  - 안내 메시지: 주문이 완료되었습니다 + 상품명, 수량, 발신자, 메시지 상세 표시
  - 확인 시 메인페이지(홈)로 자동 이동

jjw5655 committed:
fix: 주문 처리 시 product undefined 타입 오류 방지

jjw5655 committed:
chore: 사용하지 않는 login 함수의 password 파라미터 경고 제거

jjw5655 committed:
docs: 미션 2-2단계 주문하기 페이지 구현 내용으로 README.md 업데이트

greenblues1190 commented on README.md:
```diff
@@ -1,7 +1,32 @@
+  - 메시지는 반드시 입력되어야 합니다. ("메시지를 입력해주세요.")
+  - 보내는 사람 이름이 반드시 입력되어야 합니다. ("이름을 입력해주세요.")
+  - 받는 사람 이름이 반드시 입력되어야 합니다. ("이름을 입력해주세요.")
+  - 받는 사람 전화번호는 반드시 입력되고, 010-xxxx-xxxx 형식이어야 합니다. ("올바른 전화번호 형식이 아닙니다.")
```
요구사항에서 전화번호 형식이 `010-xxxx-xxxx`가 맞나요? 👀

jjw5655 replied:
참고 URL(https://kakaotech-mission2-order-step2.pages.dev/)에서 직접 입력해보니 010-1234-1234형태도 가능하고 01012341234형태도 가능하네요. 둘 다 가능하도록 코드를 추가했습니다.

혹시 입력칸에 어떤 형태의 입력이 들어가야하는지 개발자 도구로 확인할 수 있나요? 소스 탭에서 찾아봤는데 잘 안보이네요.

greenblues1190 replied:
`010-xxxx-xxxx 형식이어야 합니다` 라는 요구사항이 과제 안내 페이지에는 없었어서 여쭤보았습니다~
> 참고 URL([https://kakaotech-mission2-order-step2.pages.dev/)에서](https://kakaotech-mission2-order-step2.pages.dev/)%EC%97%90%EC%84%9C) 직접 입력해보니 010-1234-1234형태도 가능하고 01012341234형태도 가능하네요. 둘 다 가능하도록 코드를 추가했습니다.

혹시 입력칸에 어떤 형태의 입력이 들어가야하는지 개발자 도구로 확인할 수 있나요? 소스 탭에서 찾아봤는데 잘 안보이네요.
스크립트를 뜯어서 요소에 어떤 자바스크립트 로직이 들어있는지 추적하는 방법 외에는 없을 것 같아요

greenblues1190 commented on src/App.tsx:
```diff
@@ -26,7 +31,20 @@ function App() {
+            showBackButton: true,
+            showProfileButton: false,
+          };
+        }
```
switch 문으로 상단 네비게이션 바의 제목, 버튼 노출 여부등을 관리하고 계신데요, `order/:productId` 처럼 경로 깊이가 점점 더 깊어질 수록 분기문이 복잡해질 것 같아요.

`order/` 하위의 모든 경로에 대해 선물하기라는 제목을 렌더링하고 싶다면 react router의 [Outlet](https://reactrouter.com/api/components/Outlet)을 이용해서 중첩 라우트를 렌더링하는 방법에 대해 찾아보시면 좋을 것 같아요.

참고 링크: https://youngju-js.tistory.com/49

jjw5655 replied:
참고링크까지!! 감사합니다.
outlet을 활용한 중첩 라우트 구조로 변경했습니다.

greenblues1190 commented on src/App.tsx:
```diff
@@ -42,9 +60,15 @@ function App() {
+      navigate(ROUTE_MY);
+    } else {
+      // 비로그인 상태면 로그인 페이지로 이동
+      navigate(ROUTE_LOGIN, {
```
MyPage 경로는 이미 `PrivateRoute`에서 인증 여부에 따라 자동으로 리다이렉션되는데 프로필 클릭 핸들러에서도 다시 한 번 인증 체크와 분기 처리를 하고 있어요.
특별히 핸들러에서 먼저 분기해야 할 이유가 있을까요?

jjw5655 replied:
그렇네요. PrivateRoute를 적용하면서 해당 분기문은 필요없어졌군요. 삭제했습니다!

greenblues1190 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,83 @@
+      saveAuthToStorage(userData); // localStorage에 저장
+    } catch (error) {
+      throw error;
+    } finally {
```
catch 절 안에서 error을 잡아 다시 던지고 있는데 error를 잡지 않으면 기본적으로 상위 스코프로 던지기 때문에 굳이 작성 안하셔도 괜찮습니다

```suggestion
    } finally {
```

greenblues1190 commented on src/components/common/PrivateRoute.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  // 로그인되지 않은 경우 로그인 페이지로 리다이렉트
+  if (!isAuthenticated) {
+    return <Navigate to="/login" state={{ from: location.pathname }} replace />;
```
App.tsx에서 상수로 선언한 ROUTE_LOGIN를 재사용하면 이 Navigate 컴포넌트의 to에도 재사용할 수 있을 것 같아요~

```suggestion
    return <Navigate to={ROUTE_LOGIN} state={{ from: location.pathname }} replace />;
```

jjw5655 replied:
App.tsx에서 선언한 상수를 export한 후 다른 파일에서 import하여 사용하도록 구현하는게 뭔가 부자연스러워서, constants/routes.ts파일로 경로만 따로 분리한 후 여기서 import하도록 수정했습니다.

경로를 처리할 때 더 좋은 방식이 있을까요?

greenblues1190 replied:
잘 분리하신 것 같아요!

greenblues1190 commented on src/components/ranking/RankingSection.tsx:
이 파일이 조금 가독성이 떨어지는 것 같아요. 가독성 좋게 리팩터링하는 연습을 해보시면 좋을 듯 합니다

jjw5655 replied:
상수와 유틸 함수를 분리해서 리펙토링했습니다. 파라미터 관련 로직도 커스텀 훅으로 분리하면 더 깔끔할 것 같지만 일단 그냥 뒀습니다.

greenblues1190 commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,312 @@
+  const [senderError, setSenderError] = useState('');
+  const [receiverError, setReceiverError] = useState('');
+  const [phoneError, setPhoneError] = useState('');
+  const [quantityError, setQuantityError] = useState('');
```
폼 로직이 페이지 내부에 있어서 페이지의 로직을 파악하기가 어렵네요.
인풋 필드 관리를 위한 비슷한 코드가 여기저기서 반복되고 있는 것 같기도 하구요.
아래 단계로 리팩터링할 수 있을까요?

1. `useLoginForm`처럼 `useOrderForm`으로 추상화
2. `useLoginForm`과 `useOrderForm`에서 공통 코어 폼 로직을 `useForm`으로 추상화하여 두 폼 훅에서 재사용

jjw5655 replied:
바로 말씀하신대로 리펙토링했습니다.

greenblues1190 commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,312 @@
+            <CardThumbButton
+              key={card.id}
+              selected={idx === selectedCardIdx}
+              onClick={() => handleSelectCard(idx)}
```
서버에서 관리하고 있는 card의 unique key는 card.id인 것 같은데요, cardTemplates 배열의 idx를 사용하여 선택한 카드를 관리하신 이유가 있나요?

jjw5655 replied:
카드의 고유한 ID니까 이걸 그대로 사용하면 됐네요. 수정했습니다!

greenblues1190 commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,312 @@
+      <Section>
+        <div style={{ padding: 32, textAlign: 'center' }}>
+          <h2>상품을 찾을 수 없습니다.</h2>
+          <button onClick={() => navigate(-1)}>돌아가기</button>
```
NavigationBar에서는 `window.history.back()`, 여기에서는 `navigate(-1)`로 뒤로가기를 구현하고 있네요. 통일할 수 있을까요?

jjw5655 replied:
React Router를 사용하는 환경에서는 window.history.back()보다 navigate(-1)을 사용하는게 좋다고 하여 해당 방식으로 뒤로가기버튼을 통일했습니다.

greenblues1190 commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,312 @@
+
+function isValidPhoneNumber(phone: string) {
+  return /^010-\d{4}-\d{4}$/.test(phone);
+}
```
이런 유틸성 함수, 상수들은 페이지 파일에서 분리하는 것이 가독성, 유지보수에 도움이 됩니다!

jjw5655 replied:
해당 코드부분은 이전에 제안해주신 방법대로 useOrderForm.ts로 옮겼었는데요. 해당 파일에서 분리하여, 기존에 id 및 pw에 대한 validation을 수행하던 src/utils/validation.ts 파일로 옮겼습니다.

greenblues1190 commented on src/pages/MyPage.tsx:
```diff
@@ -0,0 +1,34 @@
+
+  const handleLogout = () => {
+    logout();
+    navigate('/login', { replace: true });
```
로그아웃 버튼 클릭 시 로그인 페이지로 직접 replace하셨는데, `/my` 경로는 이미 `PrivateRoute`가 감싸 인증 여부를 보고 자동으로 로그인 페이지로 replace되는 것으로 보여요.
이 라인을 제거하고 `logout()`만 호출하면 충분할 것 같습니다!

jjw5655 replied:
이 부분도 PrivateRoute덕에 필요없어졌네요. 삭제했습니다.

greenblues1190 commented on src/types/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './auth';
```
배럴 파일로 export하신 이유가 궁금해요!

jjw5655 replied:
지금은 types폴더에 auth.ts 하나밖에 없지만 앞으로 추가될 것을 고려하여 미리 배럴패턴으로 만들었습니다.
다른 폴더에서도 전부 배럴패턴으로 만들어두기도 했고 어차피 폴더 내에 여러 파일이 생기게 될 것이라, 새로운 폴더 생길때마다 그냥 배럴패턴 박고 시작하고 있습니다. 안 좋은 방식일까요?

greenblues1190 replied:
export 하는 것이 하나밖에 없어서 어떤 의도로 의사결정하셨는지 궁금하여 여쭤보았어요
안좋은 방식은 없습니다!

greenblues1190 commented on src/components/ranking/RankingSection.tsx:
```diff
@@ -114,6 +115,11 @@ const RankingSection = () => {
 
+  // 상품 클릭 시 주문 페이지로 이동
+  const handleProductClick = (product: Product) => {
+    navigate(`/order/${product.productId}`);
```
현재 템플릿 리터럴로 다이나믹 경로를 생성하고 있으신데 아무래도 직접 경로 문자열을 만들기 때문에 휴먼에러가 발생할 가능성이 있어요.

React Router의 [generatePath](https://reactrouter.com/6.30.1/utils/generate-path) 유틸을 활용하면 경로 패턴을 상수로 관리하면서 매개변수를 안전하게 주입할 수 있어 편하더라구요.

```suggestion
    import { ORDER_ROUTE } from '...'; // '/order/:productId';

    navigate(generatePath(ORDER_ROUTE), {
        productId: product.productId
    }); // '/order/:productId'의 :productId에 product.productId 삽입 
```

jjw5655 replied:
바로 generatePath 도입했습니다👍

greenblues1190 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,83 @@
+  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
+};
+
+export const useAuth = (): AuthContextType => {
+  const context = useContext(AuthContext);
+  if (context === undefined) {
+    throw new Error('useAuth must be used within an AuthProvider');
+  }
+  return context;
+};
```
인증 정보를 컨텍스트 API로 잘 관리하셨네요 💯

greenblues1190 commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,83 @@
+    };
+
+    initializeAuth();
+  }, []);
```
지금 user라는 상태를 선언하고 로컬스토리지에서 동기화하는 로직을 useEffect를 통해 구현하셨는데요,
로컬스토리지와 자동으로 동기화되는 리액트 state를 선언하는 커스텀훅을 만들면 로컬스토리지 동기화에 대한 로직을 숨길 수 있어 선언적으로 코드를 작성할 수 있을 것 같아요.

아래 형태로 user와 setUser를 사용할 수 있게끔 리팩터링해볼 수 있을까요?

```suggestion
  const [user, setUser] = useLocalStorageState<User | null>(
    '', // 로컬스토리지 key
    null // 로컬스토리지에 값이 없다면 초기값
  );
```

jjw5655 replied:
개선해봤습니다 👍

greenblues1190 commented on src/components/common/PrivateRoute.tsx:
PrivateRoute로 인증된 유저만 접근할 수 있는 페이지에 대한 라우팅 로직을 잘 구현해주셨네요 :+1

greenblues1190 reviewed (changes requested):
안녕하세요 지원님! 2단계도 잘 구현해주셨네요!!

이번에는 코멘트가 좀 많은데요, 차근차근 고민하고 진행해주시다가 어려운 점이 있으면 혼자 고민하기 보다는 편하게 댓글이나 DM으로 질문 혹은 도움 요청주세요. 함께 고민하고 같이 해결할 수 있는 부분들 찾아보도록 할게요. 

2단계 구현하시느라 수고많으셨어요~

jjw5655 committed:
feat: /order 경로 중첩 라우트 및 공통 네비게이션 레이아웃(Outlet) 적용

- /order 하위 라우트에 OrderLayout 컴포넌트와 Outlet으로 공통 NavigationBar 관리
- App.tsx 라우트 구조를 중첩 라우트 방식으로 변경
- ROUTE_ORDER 상수를 '/order'로 수정
- 분기문 제거 및 코드 구조 개선

jjw5655 committed:
refactor: 프로필 버튼 클릭 시 인증 분기 제거 및 코드 단순화

- 프로필 버튼 클릭 시 무조건 마이페이지로 이동하도록 변경
- 인증 여부 분기 및 중복 navigate 로직 제거 (리다이렉트는 PrivateRoute에서 처리)
- 프로필 버튼 인증 분기 제거로 불필요해진 코드 정리

jjw5655 committed:
refactor: 불필요한 catch 블록 제거 및 finally만 남김

- catch에서 별도 처리 없이 throw만 하던 부분을 삭제하여 코드 간결화
- 에러는 자동으로 상위로 전파됨

jjw5655 committed:
refactor: 라우트 상수 constants 폴더로 이동 및 export 구조 개선

- 라우트 경로 상수 파일을 src/constants/routes.ts로 이동
- App.tsx, PrivateRoute.tsx 등에서 import 경로를 '@/constants'로 통일
- 향후 상수 파일 추가 시 constants/index.ts에서 일괄 export 가능

jjw5655 committed:
chore: 경로 alias 설정을 '@/*' 하나로 통일

- tsconfig.app.json, vite.config.ts에서 폴더별 alias 제거
- '@/*'만 남기고 import 경로 관리 단순화

jjw5655 committed:
feat: isValidPhoneNumber 함수의 정규식 개선

- 010-1234-1234, 01012341234 두 가지 형태 모두 유효하게 처리

jjw5655 committed:
refactor: 랭킹 필터 옵션을 constants/rankingOptions.ts로 분리

- targetOptions, rankOptions를 constants/rankingOptions.ts로 이동
- constants/index.ts에서 rankingOptions 상수들 export 추가
- RankingSection.tsx에서 필터 옵션은 @/constants에서 import 사용

jjw5655 committed:
refactor: 랭킹 유틸 함수 분리 및 utils 경로 통합

- getValidValue, getValidValues, generateRankingProducts를 utils/rankingUtils.ts로 분리
- utils/index.ts에서 rankingUtils 유틸 함수 export 추가
- RankingSection.tsx에서 유틸 함수를 '@/utils'로 import
- 코드 가독성 및 유지보수성 향상

jjw5655 committed:
refactor: 폼 로직을 커스텀 훅으로 추상화 및 관심사 분리

- useOrderForm 커스텀 훅 생성하여 OrderPage 폼 로직 분리
- 공통 폼 관리를 위한 useForm 기본 훅 추가
- OrderPage에서 UI 렌더링과 폼 로직 완전 분리
- 코드 가독성 및 재사용성 향상

jjw5655 committed:
fix: /order 경로에서 NavigationBar 중복 렌더링 문제 해결

- App.tsx에서 /order 경로일 때 NavigationBar 조건부 렌더링 적용
- OrderLayout의 NavigationBar와 중복 제거하여 상단바 1개만 표시

jjw5655 committed:
refactor: 카드 선택을 배열 인덱스에서 고유 ID 기반으로 변경

- selectedCardIdx를 selectedCardId로 변경하여 card.id 사용
- handleSelectCard 함수가 카드 ID를 받도록 수정
- 배열 순서 변경에 영향받지 않는 안전한 카드 식별 방식 적용
- 서버와 클라이언트 간 카드 식별자 일관성 확보

jjw5655 committed:
refactor: 뒤로가기 동작을 navigate(-1)로 통일

- NavigationBar에서 window.history.back()을 navigate(-1)로 변경
- React Router 기반 뒤로가기 동작으로 일관성 확보
- OrderPage와 NavigationBar의 뒤로가기 방식 통일

jjw5655 committed:
refactor: isValidPhoneNumber 함수를 utils/validation으로 분리

- 전화번호 유효성 검사 함수를 utils/validation.ts로 이동
- utils/index.ts에서 모두 export하도록 수정
- 유틸 함수 재사용성 및 코드 가독성 향상

jjw5655 committed:
refactor: MyPage 로그아웃 핸들러에서 불필요한 navigate 호출 제거

- logout() 이후 navigate('/login') 호출 제거
  - 이미 PrivateRoute에 의해 자동으로 로그인 페이지 리다이렉트 처리됨
- 사용하지 않는 useNavigate관련 코드 제거
- 로그아웃 핸들러를 logout() 호출만으로 단순화

jjw5655 committed:
refactor: generatePath를 사용하여 안전한 동적 경로 생성 구현

- RankingSection에서 템플릿 리터럴 대신 generatePath 사용
- 경로 패턴 중앙 관리 및 타입 안전성 향상

jjw5655 committed:
refactor: App.tsx에서 하드코딩된 order 경로를 ROUTE_ORDER 상수로 변경

- getNavigationConfig에서 '/order/' → ROUTE_ORDER 사용
- NavigationBar 조건부 렌더링에서 '/order' → ROUTE_ORDER 사용

jjw5655 committed:
feat: useLocalStorageState 커스텀 훅 구현 및 AuthContext 리팩터링

- localStorage와 자동 동기화되는 useLocalStorageState 훅 생성
- AuthContext에서 수동 localStorage 관리 로직 제거
- 선언적 코드로 개선: setUser 호출 시 자동 저장/삭제
- 불필요한 storage 유틸리티 함수들 및 useEffect 제거
- 재사용 가능한 localStorage 상태 관리 패턴 도입

jjw5655 commented:
생각치도 못한 부분에서 고칠 점이 굉장히 많았네요. react route를 추가하면서 필요 없어진 코드들에 대한 리펙토링을 하는데 많은 도움이 되었습니다. 자세한 리뷰 감사드립니다! 수정한 사항들에 대해서도 많은 피드백 주세요. 남은 단계도 잘 부탁드립니다.

greenblues1190 commented on src/App.tsx:
```diff
@@ -9,9 +15,25 @@ import { useAuth } from '@/hooks';
 const ROUTE_NOT_FOUND = '*';
 
+// OrderLayout 컴포넌트 추가
+function OrderLayout() {
+  const navigate = useNavigate();
+  return (
+    <>
+      <NavigationBar
+        title="선물하기"
+        showBackButton={true}
+        showProfileButton={false}
+        onBackClick={() => navigate(ROUTE_HOME)}
+      />
+      <Outlet />
+    </>
+  );
+}
```
App은 리액트 웹앱의 렌더링 진입점인 만큼 다른 컴포넌트를 선언하지 않고 두는 것이 좋을 것 같습니다
OrderLayout은 다른 파일에서 선언해주세요!

greenblues1190 commented on src/constants/routes.ts:
```diff
@@ -0,0 +1,5 @@
+export const ROUTE_HOME = '/';
+export const ROUTE_LOGIN = '/login';
+export const ROUTE_MY = '/my';
+export const ROUTE_ORDER = '/order';
+export const ROUTE_NOT_FOUND = '*';
```
네임스페이스로 묶어서 다른 파일에서 사용할 때 자동완성을 편하게 하는 방법도 있을 것 같아요!
```suggestion
export const ROUTE = {
  HOME: '/',
  ...
} as const
```

greenblues1190 commented on src/utils/rankingUtils.ts:
```diff
@@ -0,0 +1,34 @@
+export const getValidValues = (options: readonly { value: string }[]) =>
+  options.map((option) => option.value);
+
+import { products } from '@/data';
+import { type Product } from '@/components/ranking/ProductCard';
```
import단이 함수 선언부 사이에 들어갔네요 👀

greenblues1190 commented on src/hooks/useForm.ts:
```diff
@@ -0,0 +1,192 @@
+  handleBlur: <K extends keyof T>(field: K) => () => void;
+}
+
+export const useForm = <T extends Record<string, any>>(config: {
+  [K in keyof T]: FieldConfig<T[K]>;
+}): UseFormReturn<T> => {
```
다양한 형태의 폼에서 재사용할 수 있도록 제네릭으로 타입 자동 추론되도록 잘 구현해주셨네요 💯

greenblues1190 commented on src/hooks/useForm.ts:
```diff
@@ -0,0 +1,192 @@
+  [K in keyof T]: FieldConfig<T[K]>;
+}): UseFormReturn<T> => {
+  // 초기 상태 생성
+  const createInitialState = () => {
+    const fields: any = {};
+    Object.keys(config).forEach((key) => {
+      fields[key] = {
+        value: config[key].defaultValue,
+        error: null,
+        touched: false,
+      };
+    });
+    return fields;
+  };
```
이 함수는 반환값이 any로 추론되어서 사용하는 곳에서 불편할 것 같아요

greenblues1190 commented on src/hooks/useForm.ts:
```diff
@@ -0,0 +1,192 @@
+  );
+
+  // 전체 폼 유효성 상태
+  const isValid =
+    Object.values(errors).every((error) => error === null) &&
+    Object.keys(config).every((key) => {
+      const fieldConfig = config[key];
+      const value = fields[key as keyof T].value;
+      return !fieldConfig.required || !!value;
+    });
```
가독성을 위해 긴 조건들은 변수로 추상화하여 변수명으로 의미를 드러내도록 하면 유지보수에 좋을 것 같아요

예시
```suggestion
  const isValid = !hasErrors || !missingRequired;
```

greenblues1190 commented on src/hooks/useLocalStorageState.ts:
**[참고]** 깔끔하게 추상화되었네요. 이제 로컬스토리지와 리액트 상태 동기화에 대한 내부로직에 대한 관심은 이 훅에 맡기고 편하게 사용할 수 있을 것 같습니다.
(만약 필요하다면) [스토리지 이벤트](https://blog.shiren.dev/ko/post/2023-04-18)를 구독하여 고도화할 수도 있겠네요 👍

greenblues1190 reviewed (approved):
안녕하세요 지원님! 피드백이 많았음에도 잘 반영해주셨네요! 코드가 더 깔끔해지고 변경에 유연한 서비스가 된 것 같습니다! 다음 단계도 화이팅입니다!
